### PR TITLE
Metrics for Table Disabled and Consumption Paused

### PR DESCRIPTION
**Pull request by suddendust**
Tests:

Screenshot 2023-12-01 at 1 58 44 AM

Exported Metrics:
Screenshot 2023-12-01 at 1 59 05 AM

Screenshot 2023-12-01 at 2 00 02 AM

### CODE REVIEW

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 1]
```diff
@@ -871,8 +872,8 @@ public void ensureAllPartitionsConsuming(TableConfig tableConfig, StreamConfig s
       boolean offsetsHaveToChange = offsetCriteria != null;
       if (isTableEnabled && !isTablePaused) {
         List<PartitionGroupConsumptionStatus> currentPartitionGroupConsumptionStatusList =
-            offsetsHaveToChange
-                ? Collections.emptyList() // offsets from metadata are not valid anymore; fetch for all partitions
+            offsetsHaveToChange ? Collections.emptyList()
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:875]: (minor) This is not as readable. Suggest not reformatting unrelated code

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 2]
```diff
@@ -1531,9 +1530,10 @@ public PauseStatus pauseConsumption(String tableNameWithType) {
     IdealState updatedIdealState = updatePauseStatusInIdealState(tableNameWithType, true);
     Set<String> consumingSegments = findConsumingSegments(updatedIdealState);
     sendForceCommitMessageToServers(tableNameWithType, consumingSegments);
-    return new PauseStatus(true, consumingSegments, consumingSegments.isEmpty() ? null : "Pause flag is set."
-        + " Consuming segments are being committed."
-        + " Use /pauseStatus endpoint in a few moments to check if all consuming segments have been committed.");
+    _controllerMetrics.setValueOfTableGauge(tableNameWithType, ControllerGauge.TABLE_CONSUMPTION_PAUSED, 1);
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:1533]: This won't capture the paused table when controller is restarted. We need to set this gauge when a segment commits, or in `ensureAllPartitionsConsuming()`

Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:1533]: `ensureAllPartitionsConsuming` is also in the controller, so if it restarts, will it capture it?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 1 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java]: Reverted this change.

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml [hunk 3]
```diff
@@ -204,6 +204,18 @@ rules:
   cache: true
   labels:
     version: "$2"
+- pattern: "\"org.apache.pinot.common.metrics\"<type=\"ControllerMetrics\", name=\"pinot.controller.tableConsumptionPaused.([^\\.]*?)_(OFFLINE|REALTIME)\"><>(\\w+)"
+  name: "pinot_controller_tableConsumptionPaused_$3"
+  cache: true
+  labels:
+    tableName: "$1"
+    tableType: "$2"
```
Comment (zhtaoxiang) [inline docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml:212]: nit: I think table consumption only applies to realtime table?

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java [hunk 4]
```diff
@@ -61,6 +61,7 @@
 public class SegmentStatusChecker extends ControllerPeriodicTask<SegmentStatusChecker.Context> {
   private static final Logger LOGGER = LoggerFactory.getLogger(SegmentStatusChecker.class);
   private static final int MAX_OFFLINE_SEGMENTS_TO_LOG = 5;
+  private static final String TABLE_CONSUMPTION_PAUSED = "isTablePaused";
```
Comment (zhtaoxiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:64]: Can we use `isTableConsumptionPaused` to make it easy to read?

File: docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml (hunk 3 above)
Comment (suddendust) [reply docker/images/pinot/etc/jmx_prometheus_javaagent/configs/controller.yml:212]: Maybe in future we can add an ability to pause consumption in offline tables too? I think we can keep this label.

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java (hunk 4 above)
Comment (suddendust) [reply pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java]: Actually, this is the key that is set is the table idealstate when a table is paused. So will have to use it to for lookup.

**Review by zhtaoxiang — APPROVED** [review]
LGTM

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java [hunk 5]
```diff
@@ -61,6 +61,7 @@
 public class SegmentStatusChecker extends ControllerPeriodicTask<SegmentStatusChecker.Context> {
   private static final Logger LOGGER = LoggerFactory.getLogger(SegmentStatusChecker.class);
   private static final int MAX_OFFLINE_SEGMENTS_TO_LOG = 5;
+  private static final String IS_TABLE_CONSUMPTION_PAUSED = "isTablePaused";
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:64]: Let's reuse `PinotLLCRealtimeSegmentManager.IS_TABLE_PAUSED` (make it public)

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java [hunk 6]
```diff
@@ -369,5 +396,7 @@ public static final class Context {
     private int _offlineTableCount;
     private int _disabledTableCount;
     private Set<String> _processedTables = new HashSet<>();
+    private Set<String> _disabledTables = new HashSet<>();
```
Comment (Jackie-Jiang) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/SegmentStatusChecker.java:399]: Since we already track this set, no need to track `_disabledTableCount`

**Review by Jackie-Jiang — APPROVED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java [hunk 7]
```diff
@@ -129,14 +129,14 @@
  * TODO: migrate code in this class to other places for better readability
  */
 public class PinotLLCRealtimeSegmentManager {
+
+  // simple field in Ideal State representing pause status for the table
+  public static final String IS_TABLE_PAUSED = "isTablePaused";
```
Comment (walterddr) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:134]: what was this change for?

**Review by walterddr — COMMENTED** [review]
lgtm. minor comment

File: pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java (hunk 7 above)
Comment (walterddr) [inline pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java:134]: instead of exposing a private static field. we can probably expose the isTablePause method as public static. but can be follow up
